src/librustc_codegen_ssa/mir/analyze.rs

```diff
@@ -261,9 +131,67 @@ impl<'mir, 'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> Visitor<'tcx>
         self.super_terminator(terminator, location);
     }
 
-    fn visit_place(&mut self, place: &mir::Place<'tcx>, context: PlaceContext, location: Location) {
+    fn visit_place(
+        &mut self,
+        place: &mir::Place<'tcx>,
+        mut context: PlaceContext,
+        location: Location,
+    ) {
         debug!("visit_place(place={:?}, context={:?})", place, context);
-        self.process_place(&place.as_ref(), context, location);
+
+        // Non-uses do not force locals onto the stack.
+        if !context.is_use() {
+            return;
+        }
+
+        let mir::Place { local, projection } = *place;
+
+        // Reads from ZSTs do not require memory accesses and do not count when determining what
+        // needs to live on the stack.
+        if is_consume(context) {
+            let ty = place.ty(self.fx.mir, self.fx.cx.tcx()).ty;
+            let ty = self.fx.monomorphize(&ty);
+            let span = self.fx.mir.local_decls[local].source_info.span;
+            if self.fx.cx.spanned_layout_of(ty, span).is_zst() {
+                return;
+            }
+        }
+
+        let mut has_disqualifying_projection = false;
+
+        let mut projection: &[_] = projection.as_ref();
+        while let [ref proj_base @ .., elem] = *projection {
+            projection = proj_base;
+            self.super_projection_elem(local, proj_base, elem, context, location);
+
+            // Projections like `(*x)[12]` are allowed but not `*(x[12])`, since a `Deref` of a
+            // local acts like a `Copy` of that local.
+            if let mir::PlaceElem::Deref = elem {
+                has_disqualifying_projection = false;
+                context = PlaceContext::NonMutatingUse(NonMutatingUseContext::Copy);
+                continue;
+            }
+
+            // Ignoring `Deref`s, the only allowed projections are reads of scalar fields.
+            if is_consume(context) && matches!(elem, mir::ProjectionElem::Field(..)) {
+                let base_ty = mir::Place::ty_from(local, proj_base, self.fx.mir, self.fx.cx.tcx());
+                let base_ty = self.fx.monomorphize(&base_ty);
+                let span = self.fx.mir.local_decls[local].source_info.span;
+                let layout = self.fx.cx.spanned_layout_of(base_ty.ty, span);
+
+                if !ty_requires_alloca(self.fx, layout) {
+                    continue;
+                }
+            }
+
+            has_disqualifying_projection = true;
+        }
```

Comment:
Looking more at this loop, I would split it into several parts:
* a loop to just call `super_projection_elem` uniformly on all the elems, to get it out of the way (it only serves to visit the index local of an `Index` projection, with a `Copy context)
* find the *first* `Deref`, if there is one, e.g. with  `.position(|elem| matches!(elem, mir::PlaceElem::Deref))`
* loop over the "direct projections" (i.e. before any `Deref`, if any) and do `self.not_ssa(local); break;` at the first sign of trouble

I mostly want this for clarity, but that last part has some promise for making this more efficient, so here's a few more ideas:
* `context` wouldn't change across it, meaning `if is_consume(context)` can be placed around the whole loop
* there could be a quick check that all the elements are fields, to avoid time even looking at the layouts
* we probably should check all the field projections (like today / this PR) but if we go forward, from the `local` to the innermost field, we should almost always hit the non-SSA layout first, making the rejection cheaper
* we could call `visit_local` before the loop, and skip the loop if `non_ssa_locals` already contains the local (see also the comment at the top of this file, which would make this more useful)
* if we keep the monomorphized `TyAndLayout` of the local (which we always have to compute anyway) around, we could start from it and just go `layout = layout.field(self.fx, i)` at every step instead of computing the whole type and monomorphizing every time